### PR TITLE
New Ore Silo UI

### PR DESCRIPTION
**Pull request by intercepti0n**
UI

## About The Pull Request

Made a new UI for Ore Silo, tweaked `MaterialAccessBar` component a bit to properly display pop-ups over “wrapped” elements and actually wrap.

## Why It's Good For The Game

Ore Silo is one of the last machines that use browser on TG, we're almost there!

Also, it helps @JohnFulpWillard to bring https://hackmd.io/XLt5MoRvRxuhFbwtk4VAUA HackMD document to a closure.

## Changelog
:cl:
refactor: Refactored Ore Silo Ui.
/:cl:

### CODE REVIEW

File: code/modules/mining/machine_silo.dm [hunk 1]
```diff
@@ -77,93 +77,74 @@ GLOBAL_LIST_EMPTY(silo_access_logs)
...
+			"rawMaterials" = entry.get_raw_materials(""),
+			"machineName" = entry.machine_name,
+			"areaName" = entry.area_name,
+			"action" = entry.action,
+			"amount" = entry.amount,
+			"time" = entry.timestamp,
+			"noun" = entry.noun
+		)
+		LAZYADD(logs_data, list(log_data))
+	LAZYSET(data, "logs", logs_data)
```
Comment (Mothblocks) [inline code/modules/mining/machine_silo.dm:125]: Why bother with lazy lists? Just makes this code more complicated.

File: code/modules/mining/machine_silo.dm [hunk 2]
```diff
@@ -77,93 +77,74 @@ GLOBAL_LIST_EMPTY(silo_access_logs)
...
+		if("remove")
+			var/datum/component/remote_materials/remote = locate(params["ref"]) in ore_connected_machines
+			remote?.disconnect_from(src)
+			return TRUE
+		if("hold")
+			var/datum/component/remote_materials/remote = locate(params["ref"]) in ore_connected_machines
+			remote?.toggle_holding()
+			return TRUE
+		if("eject")
+			var/datum/material/ejecting = locate(params["ref"])
```
Comment (Mothblocks) [inline code/modules/mining/machine_silo.dm:144]: You need to validate that this is actually a /datum/material.

Also is this pointing to a singleton? I'm not up to date with materials code. If it's owned by us in any way you should validate that too.

File: tgui/packages/tgui/interfaces/OreSilo.tsx [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+          }}
+        >
+          <Icon name="trash-can" />
+        </Box>
+      </Tooltip>
+    </Box>
+  );
+};
+
+interface LogProps {
```
Comment (Mothblocks) [inline tgui/packages/tgui/interfaces/OreSilo.tsx:174]: I haven't seen tgui code in a while but we don't usually use interface, do we? Should be `type LogProps = {` unless I'm mistaken.

Comment (Mothblocks) [reply tgui/packages/tgui/interfaces/OreSilo.tsx:174]: Here and everywhere else

File: tgui/packages/tgui/interfaces/OreSilo.tsx [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+  machineName: string;
+  areaName: string;
+  action: string;
+  amount: number;
+  time: string;
+  noun: string;
+}
+
+type OreSiloData = {
+  SHEET_MATERIAL_AMOUNT: number;
```
Comment (Mothblocks) [inline tgui/packages/tgui/interfaces/OreSilo.tsx:40]: This should just be cased normally, looks weird like this

File: code/modules/mining/machine_silo.dm [hunk 5]
```diff
@@ -77,93 +77,74 @@ GLOBAL_LIST_EMPTY(silo_access_logs)
...
+	var/list/connected_data
+	for(var/datum/component/remote_materials/remote as anything in ore_connected_machines)
+		var/atom/parent = remote.parent
+		var/icon/parent_icon = icon(initial(parent.icon), initial(parent.icon_state), frame = 1)
+		var/list/remote_data = list(
+			"ref" = REF(remote),
+			"icon" = icon2base64(parent_icon),
+			"name" = parent.name,
+			"onHold" = !!holds[remote],
+			"location" = get_area_name(parent, TRUE)
```
Comment (Mothblocks) [inline code/modules/mining/machine_silo.dm:108]: Named param for true

**Review by Mothblocks — CHANGES_REQUESTED** [review]

File: tgui/packages/tgui/interfaces/OreSilo.tsx (hunk 4 above)
Comment (intercepti0n) [reply tgui/packages/tgui/interfaces/OreSilo.tsx:40]: This parameter is being handled by `/datum/component/material_container` in `ui_static_data` proc, I can't change it, unfortunately.

File: tgui/packages/tgui/interfaces/OreSilo.tsx (hunk 3 above)
Comment (intercepti0n) [reply tgui/packages/tgui/interfaces/OreSilo.tsx:174]: Sure, but what's the difference between those?

File: code/modules/mining/machine_silo.dm (hunk 1 above)
Comment (optimumtact) [reply code/modules/mining/machine_silo.dm:125]: it's free memory

File: code/modules/mining/machine_silo.dm (hunk 2 above)
Comment (intercepti0n) [reply code/modules/mining/machine_silo.dm:144]: Sending material refs is handled by `/datum/component/remote_materials` in `ui_data` that guarantees that all of them will be valid. Code above was taken from `/obj/machinery/rnd/production`, commentary there says “SAFETY: eject_sheets checks for valid mats”.

File: tgui/packages/tgui/interfaces/OreSilo.tsx (hunk 4 above)
Comment (intercepti0n) [reply tgui/packages/tgui/interfaces/OreSilo.tsx:40]: Oh, wait. I'm dumb, huh.

File: code/modules/mining/machine_silo.dm (hunk 2 above)
Comment (intercepti0n) [reply code/modules/mining/machine_silo.dm:144]: Uh, I can assume that material will be probably valid. 😅

**Review by JohnFulpWillard — COMMENTED** [review]
Remove ``INTERACT_MACHINE_SET_MACHINE`` from the silo's interaction flags please

File: tgui/packages/tgui/interfaces/OreSilo.tsx [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+};
+
+export const OreSilo = (props: any, context: any) => {
+  const { act, data } = useBackend<OreSiloData>();
+  const { SHEET_MATERIAL_AMOUNT, machines, logs } = data;
+
+  const [currentTab, setCurrentTab] = useState(0);
+
+  return (
+    <Window title="Ore Silo" width={400} height={600}>
```
Comment (SyncIt21) [inline tgui/packages/tgui/interfaces/OreSilo.tsx:53]: Increase the width of this window so the material bar isn wrapped around the window, it looks ugly and unaligned, just make the same size as your techfab window

**Review by SyncIt21 — CHANGES_REQUESTED** [review]

File: tgui/packages/tgui/interfaces/OreSilo.tsx (hunk 6 above)
Comment (JohnFulpWillard) [reply tgui/packages/tgui/interfaces/OreSilo.tsx:53]: i thought it looked fine

Comment (SyncIt21) [reply tgui/packages/tgui/interfaces/OreSilo.tsx:53]: Nothing wrong with it true,  but people are used to seeing the material bar displayed in a single line in all machines.
No need to break that continuity now

File: code/modules/mining/machine_silo.dm (hunk 2 above)
Comment (intercepti0n) [reply code/modules/mining/machine_silo.dm:146]: Sure, added some.

**Pull request merged**